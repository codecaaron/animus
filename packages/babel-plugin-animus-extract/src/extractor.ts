import type { NodePath, types as BabelTypes } from '@babel/core';
import type { CallExpression } from '@babel/types';

import { generateCSS } from './css-generator';
import { createRuntimeReplacement } from './transformer';
import type { AnimusExtractState, ComponentData, ExtractedChainCall } from './types';
import { generateComponentId } from './utils';

export function isAnimusBuilder(
  path: NodePath<CallExpression>,
  state: AnimusExtractState
): boolean {
  const { types: t } = state;
  const {callee} = path.node;

  // Direct call: animus.styles()
  if (
    t.isMemberExpression(callee) &&
    t.isIdentifier(callee.object) &&
    (callee.object.name === 'animus' || callee.object.name === state.animusBinding)
  ) {
    return true;
  }

  // Chained call: animus.styles().states()
  if (t.isMemberExpression(callee) && t.isCallExpression(callee.object)) {
    const objectPath = path.get('callee.object') as NodePath<CallExpression>;
    return isAnimusBuilder(objectPath, state);
  }

  return false;
}

function isAnimusChainCall(path: NodePath<CallExpression>, state: AnimusExtractState): boolean {
  const { types: t } = state;
  const {callee} = path.node;

  return t.isMemberExpression(callee) &&
    (isAnimusBuilder(path, state) ||
     (t.isCallExpression(callee.object) && isAnimusChainCall(path.get('callee.object') as NodePath<CallExpression>, state)));
}

export function extractAnimusComponent(
  path: NodePath<CallExpression>,
  state: AnimusExtractState
): void {
  const chain = extractChain(path, state);
  if (chain.length === 0) return;

  const componentData = processChain(chain, state);

  // Generate CSS for this component
  const css = generateCSS(componentData, state);
  if (css) {
    state.cssOutput.set(componentData.id, css);
  }

  // Store component data for JSX transformation
  if (componentData.name) {
    state.animusComponents.set(componentData.name, componentData);
  }

  // Replace the original expression with runtime version
  const replacement = createRuntimeReplacement(componentData, state);
  path.replaceWith(replacement);
}

function extractChain(
  path: NodePath<CallExpression>,
  state: AnimusExtractState
): ExtractedChainCall[] {
  const chain: ExtractedChainCall[] = [];
  let current: NodePath<CallExpression> | null = path;

  while (current && isAnimusChainCall(current, state)) {
    const { types: t } = state;
    const {callee} = current.node;

    if (t.isMemberExpression(callee) && t.isIdentifier(callee.property)) {
      const methodName = callee.property.name;
      const args = current.node.arguments;

      chain.unshift({ methodName, args });

      if (t.isCallExpression(callee.object)) {
        current = current.get('callee.object') as NodePath<CallExpression>;
      } else {
        current = null;
      }
    } else {
      current = null;
    }
  }

  return chain;
}

function processChain(
  chain: ExtractedChainCall[],
  state: AnimusExtractState
): ComponentData {
  const { types: t } = state;
  const componentData: ComponentData = {
    id: generateComponentId(),
  };

  for (const { methodName, args } of chain) {
    switch (methodName) {
      case 'styles':
        if (args[0] && t.isObjectExpression(args[0])) {
          componentData.styles = extractObjectLiteral(args[0], state);
        }
        break;

      case 'states':
        if (args[0] && t.isObjectExpression(args[0])) {
          componentData.states = extractObjectLiteral(args[0], state);
        }
        break;

      case 'variants':
        if (args[0] && t.isObjectExpression(args[0])) {
          componentData.variants = extractObjectLiteral(args[0], state);
        }
        break;

      case 'groups':
        if (args[0] && t.isObjectExpression(args[0])) {
          componentData.groups = extractObjectLiteral(args[0], state);
        }
        break;

      case 'props':
        if (args[0] && t.isObjectExpression(args[0])) {
          componentData.props = extractObjectLiteral(args[0], state);
        }
        break;

      case 'asElement':
        if (args[0] && t.isStringLiteral(args[0])) {
          componentData.element = args[0].value;
        }
        break;

      case 'displayName':
        if (args[0] && t.isStringLiteral(args[0])) {
          componentData.name = args[0].value;
        }
        break;
    }
  }

  return componentData;
}

function extractObjectLiteral(
  node: BabelTypes.ObjectExpression,
  state: AnimusExtractState
): Record<string, any> {
  const { types: t } = state;
  const result: Record<string, any> = {};

  for (const prop of node.properties) {
    if (t.isObjectProperty(prop) || t.isObjectMethod(prop)) {
      const key = getPropertyKey(prop, state);
      if (key && t.isObjectProperty(prop)) {
        const value = extractValue(prop.value, state);
        if (value !== undefined) {
          result[key] = value;
        }
      }
    }
  }

  return result;
}

function getPropertyKey(
  prop: BabelTypes.ObjectProperty | BabelTypes.ObjectMethod,
  state: AnimusExtractState
): string | null {
  const { types: t } = state;

  if (t.isIdentifier(prop.key)) {
    return prop.key.name;
  } else if (t.isStringLiteral(prop.key)) {
    return prop.key.value;
  }

  return null;
}

function extractValue(node: BabelTypes.Node, state: AnimusExtractState): any {
  const { types: t } = state;

  if (t.isStringLiteral(node)) {
    return node.value;
  } else if (t.isNumericLiteral(node)) {
    return node.value;
  } else if (t.isBooleanLiteral(node)) {
    return node.value;
  } else if (t.isObjectExpression(node)) {
    return extractObjectLiteral(node, state);
  } else if (t.isArrayExpression(node)) {
    return node.elements.map(el => el ? extractValue(el, state) : null);
  } else if (t.isUnaryExpression(node) && node.operator === '-' && t.isNumericLiteral(node.argument)) {
    return -node.argument.value;
  }

  // For dynamic values, return undefined to skip extraction
  return undefined;
}
