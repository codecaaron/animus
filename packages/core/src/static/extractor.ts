import * as parser from '@babel/parser';
// Handle the babel/traverse CommonJS export issue
// @ts-ignore - babel/traverse has complex module exports
import traverseDefault from '@babel/traverse';
import * as t from '@babel/types';

const traverse = (traverseDefault as any).default || traverseDefault;

import type { NodePath } from '@babel/traverse';

/**
 * Extracted style data from an Animus component
 */
export interface ExtractedStyles {
  componentName?: string;
  baseStyles?: Record<string, any>;
  variants?: Record<string, any> | Record<string, any>[];
  states?: Record<string, any>;
  groups?: string[];
  props?: Record<string, any>;
}

/**
 * Extract styles from Animus method chains in code
 */
export function extractStylesFromCode(code: string): ExtractedStyles[] {
  const ast = parser.parse(code, {
    sourceType: 'module',
    plugins: ['typescript', 'jsx'],
  });

  const extractedComponents: ExtractedStyles[] = [];

  traverse(ast as any, {
    CallExpression(path: NodePath<t.CallExpression>) {
      // Look for .styles() calls
      if (
        t.isMemberExpression(path.node.callee) &&
        t.isIdentifier(path.node.callee.property, { name: 'styles' })
      ) {
        const extracted: ExtractedStyles = {};

        // Try to get component name from variable declaration
        const varDeclarator = path.findParent((p: NodePath) =>
          p.isVariableDeclarator()
        );
        if (varDeclarator && t.isVariableDeclarator(varDeclarator.node)) {
          if (t.isIdentifier(varDeclarator.node.id)) {
            extracted.componentName = varDeclarator.node.id.name;
          }
        }

        // Extract styles object from first argument
        const stylesArg = path.node.arguments[0];
        if (t.isObjectExpression(stylesArg)) {
          extracted.baseStyles = extractObjectLiteral(stylesArg);
        }

        // Look for chained method calls by traversing up the AST
        // We need to find the full chain starting from .styles()
        const chainedCalls: any[] = [];
        let currentPath: any = path;

        // Traverse up to find all chained calls
        while (currentPath) {
          if (
            t.isCallExpression(currentPath.node) &&
            t.isMemberExpression(currentPath.node.callee) &&
            t.isIdentifier(currentPath.node.callee.property)
          ) {
            chainedCalls.push({
              method: currentPath.node.callee.property.name,
              args: currentPath.node.arguments,
            });
          }

          // Check if parent is also a call expression (chained)
          if (
            currentPath.parentPath &&
            t.isMemberExpression(currentPath.parentPath.node) &&
            currentPath.parentPath.parentPath &&
            t.isCallExpression(currentPath.parentPath.parentPath.node)
          ) {
            currentPath = currentPath.parentPath.parentPath;
          } else {
            break;
          }
        }

        // Process chained calls in order
        const variantCalls: any[] = [];

        for (const call of chainedCalls) {
          if (call.method === 'variant' && call.args[0]) {
            const variantArg = call.args[0];
            if (t.isObjectExpression(variantArg)) {
              variantCalls.push(extractObjectLiteral(variantArg));
            }
          } else if (call.method === 'states' && call.args[0]) {
            const statesArg = call.args[0];
            if (t.isObjectExpression(statesArg)) {
              extracted.states = extractObjectLiteral(statesArg);
            }
          } else if (call.method === 'groups' && call.args[0]) {
            const groupsArg = call.args[0];
            if (t.isObjectExpression(groupsArg)) {
              const groups = extractObjectLiteral(groupsArg);
              extracted.groups = Object.keys(groups).filter(
                (key) => groups[key]
              );
            }
          } else if (call.method === 'props' && call.args[0]) {
            const propsArg = call.args[0];
            if (t.isObjectExpression(propsArg)) {
              extracted.props = extractObjectLiteral(propsArg);
            }
          }
        }

        // Store variants - if multiple, keep as array; if single, flatten
        if (variantCalls.length > 1) {
          extracted.variants = variantCalls;
        } else if (variantCalls.length === 1) {
          extracted.variants = variantCalls[0];
        }

        extractedComponents.push(extracted);
      }
    },
  });

  return extractedComponents;
}

/**
 * Extract a literal object from an AST ObjectExpression
 */
function extractObjectLiteral(node: t.ObjectExpression): Record<string, any> {
  const result: Record<string, any> = {};

  for (const prop of node.properties) {
    if (t.isObjectProperty(prop) && !prop.computed) {
      const key = t.isIdentifier(prop.key)
        ? prop.key.name
        : t.isStringLiteral(prop.key)
          ? prop.key.value
          : null;

      if (key) {
        const value = extractValue(prop.value);
        if (value !== undefined) {
          result[key] = value;
        }
      }
    }
  }

  return result;
}

/**
 * Extract a literal value from an AST node
 */
function extractValue(node: t.Node): any {
  if (t.isStringLiteral(node)) {
    return node.value;
  } else if (t.isNumericLiteral(node)) {
    return node.value;
  } else if (t.isBooleanLiteral(node)) {
    return node.value;
  } else if (t.isNullLiteral(node)) {
    return null;
  } else if (t.isObjectExpression(node)) {
    return extractObjectLiteral(node);
  } else if (t.isArrayExpression(node)) {
    return node.elements
      .filter((el): el is t.Expression | t.SpreadElement => el !== null)
      .map((el) => (t.isExpression(el) ? extractValue(el) : undefined))
      .filter((v) => v !== undefined);
  } else if (t.isTemplateLiteral(node) && node.expressions.length === 0) {
    // Simple template literal with no expressions
    return node.quasis[0].value.cooked;
  }

  // Return undefined for non-literal expressions
  return undefined;
}
