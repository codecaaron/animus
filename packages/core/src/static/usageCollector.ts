import * as parser from '@babel/parser';
import * as t from '@babel/types';

// Handle the babel/traverse CommonJS export issue
// @ts-ignore - babel/traverse has complex module exports
import traverseDefault from '@babel/traverse';
const traverse = (traverseDefault as any).default || traverseDefault;
import type { NodePath } from '@babel/traverse';

/**
 * Usage value with optional breakpoint information
 */
export interface UsageValue {
  value: any;
  breakpoint?: string; // '_' for base, 'sm', 'md', etc.
}

/**
 * Map of component usage - tracks which prop values are actually used
 */
export interface UsageMap {
  [componentName: string]: {
    [propName: string]: Set<string>; // Set of "value:breakpoint" strings
  };
}

/**
 * Component reference found in JSX
 */
export interface ComponentUsage {
  componentName: string;
  props: Record<string, any>;
  location?: {
    line: number;
    column: number;
  };
}

/**
 * Extract component usage from JSX in code
 */
export function extractComponentUsage(code: string): ComponentUsage[] {
  const ast = parser.parse(code, {
    sourceType: 'module',
    plugins: ['typescript', 'jsx'],
  });

  const usages: ComponentUsage[] = [];

  traverse(ast as any, {
    JSXOpeningElement(path: NodePath<t.JSXOpeningElement>) {
      // Get component name
      let componentName: string | null = null;

      if (t.isJSXIdentifier(path.node.name)) {
        // Simple component: <Box />
        const name = path.node.name.name;
        // Only track capitalized components (not HTML elements)
        if (name[0] === name[0].toUpperCase()) {
          componentName = name;
        }
      } else if (
        t.isJSXMemberExpression(path.node.name) &&
        t.isJSXIdentifier(path.node.name.object) &&
        t.isJSXIdentifier(path.node.name.property)
      ) {
        // Compound component: <Layout.Header />
        componentName = `${path.node.name.object.name}.${path.node.name.property.name}`;
      }

      if (!componentName) return;

      // Extract props
      const props: Record<string, any> = {};

      for (const attr of path.node.attributes) {
        if (t.isJSXAttribute(attr) && t.isJSXIdentifier(attr.name)) {
          const propName = attr.name.name;
          let propValue: any = true; // default for boolean props like <Box disabled />

          if (attr.value) {
            if (t.isJSXExpressionContainer(attr.value)) {
              // {expression}
              propValue = extractJSXExpressionValue(attr.value.expression);
            } else if (t.isStringLiteral(attr.value)) {
              // "string"
              propValue = attr.value.value;
            }
          }

          if (propValue !== undefined) {
            props[propName] = propValue;
          }
        }
        // TODO: Handle spread props {...props}
      }

      const usage: ComponentUsage = {
        componentName,
        props,
      };

      // Add location for debugging
      if (path.node.loc) {
        usage.location = {
          line: path.node.loc.start.line,
          column: path.node.loc.start.column,
        };
      }

      usages.push(usage);
    },
  });

  return usages;
}

/**
 * Extract value from JSX expression container
 */
function extractJSXExpressionValue(
  node: t.Expression | t.JSXEmptyExpression
): any {
  if (t.isStringLiteral(node)) {
    return node.value;
  } else if (t.isNumericLiteral(node)) {
    return node.value;
  } else if (t.isBooleanLiteral(node)) {
    return node.value;
  } else if (t.isNullLiteral(node)) {
    return null;
  } else if (t.isArrayExpression(node)) {
    // Handle responsive arrays: p={[1, 2, 3]}
    const values = [];
    for (const el of node.elements) {
      if (el && t.isExpression(el)) {
        const value = extractJSXExpressionValue(el);
        if (value !== undefined) {
          values.push(value);
        }
      }
    }
    return values.length > 0 ? values : undefined;
  } else if (t.isObjectExpression(node)) {
    // Handle responsive objects: p={{ _: 1, sm: 2 }}
    const obj: Record<string, any> = {};
    for (const prop of node.properties) {
      if (t.isObjectProperty(prop) && !prop.computed) {
        const key = t.isIdentifier(prop.key)
          ? prop.key.name
          : t.isStringLiteral(prop.key)
            ? prop.key.value
            : null;

        if (key && t.isExpression(prop.value)) {
          const value = extractJSXExpressionValue(prop.value);
          if (value !== undefined) {
            obj[key] = value;
          }
        }
      }
    }
    return Object.keys(obj).length > 0 ? obj : undefined;
  }

  // Return undefined for non-literal expressions
  return undefined;
}

/**
 * Build usage map from component usages
 */
export function buildUsageMap(usages: ComponentUsage[]): UsageMap {
  const map: UsageMap = {};

  // Breakpoint order from propertyMappings
  const breakpointOrder = ['_', 'xs', 'sm', 'md', 'lg', 'xl'];

  for (const usage of usages) {
    if (!map[usage.componentName]) {
      map[usage.componentName] = {};
    }

    for (const [propName, propValue] of Object.entries(usage.props)) {
      if (!map[usage.componentName][propName]) {
        map[usage.componentName][propName] = new Set();
      }

      // Handle different value types
      if (Array.isArray(propValue)) {
        // Responsive array: map to breakpoints
        propValue.forEach((v, index) => {
          if (v !== undefined && index < breakpointOrder.length) {
            const breakpoint = breakpointOrder[index];
            map[usage.componentName][propName].add(`${v}:${breakpoint}`);
          }
        });
      } else if (
        propValue &&
        typeof propValue === 'object' &&
        !Array.isArray(propValue)
      ) {
        // Responsive object: extract breakpoint from keys
        Object.entries(propValue).forEach(([breakpoint, v]) => {
          if (v !== undefined) {
            map[usage.componentName][propName].add(`${v}:${breakpoint}`);
          }
        });
      } else {
        // Single value - base breakpoint
        map[usage.componentName][propName].add(`${propValue}:_`);
      }
    }
  }

  return map;
}

/**
 * Merge multiple usage maps
 */
export function mergeUsageMaps(...maps: UsageMap[]): UsageMap {
  const merged: UsageMap = {};

  for (const map of maps) {
    for (const [componentName, props] of Object.entries(map)) {
      if (!merged[componentName]) {
        merged[componentName] = {};
      }

      for (const [propName, values] of Object.entries(props)) {
        if (!merged[componentName][propName]) {
          merged[componentName][propName] = new Set();
        }

        values.forEach((v) => merged[componentName][propName].add(v));
      }
    }
  }

  return merged;
}
