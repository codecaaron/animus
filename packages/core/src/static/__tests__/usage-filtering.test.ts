import { extractStylesFromCode } from '../extractor';
import { CSSGenerator } from '../generator';
import { buildUsageMap, extractComponentUsage } from '../usageCollector';
import { groupDefinitions, testTheme } from './testConfig';

describe('Usage-based atomic CSS filtering', () => {
  it('should only generate utilities for actually used prop values', () => {
    // Component definition with space group enabled
    const componentCode = `
      const Box = animus
        .styles({
          display: 'block',
          position: 'relative'
        })
        .groups({ space: true })
        .asElement('div');
    `;

    // Usage code showing which props are actually used
    const usageCode = `
      function App() {
        return (
          <>
            <Box p={4} />
            <Box mx={2} />
            <Box p={[1, 2]} />
            <Box m={0} />
          </>
        );
      }
    `;

    // Extract component definition
    const extracted = extractStylesFromCode(componentCode);
    expect(extracted).toHaveLength(1);

    // Extract usage
    const usages = extractComponentUsage(usageCode);
    const usageMap = buildUsageMap(usages);

    // Generate CSS without usage map (baseline - no utilities without usage data)
    const generator = new CSSGenerator({ prefix: 'animus' });
    const withoutUsage = generator.generateFromExtracted(
      extracted[0],
      groupDefinitions,
      testTheme
    );

    // Generate CSS with usage map (filtered - only used utilities)
    const withUsage = generator.generateFromExtracted(
      extracted[0],
      groupDefinitions,
      testTheme,
      usageMap
    );

    // Verify usage map contains expected values
    expect(usageMap.Box).toBeDefined();
    expect(usageMap.Box.p).toEqual(new Set(['4:_', '1:_', '2:xs']));
    expect(usageMap.Box.mx).toEqual(new Set(['2:_']));
    expect(usageMap.Box.m).toEqual(new Set(['0:_']));

    // Snapshot CSS outputs
    expect({
      withoutUsageFiltering: withoutUsage.css,
      withUsageFiltering: withUsage.css,
    }).toMatchSnapshot();
  });

  it('should handle responsive values in usage tracking', () => {
    const componentCode = `
      const Card = animus
        .styles({
          backgroundColor: 'white',
          borderRadius: '8px'
        })
        .groups({ space: true })
        .asElement('div');
    `;

    const usageCode = `
      function Layout() {
        return (
          <Card p={{ _: 2, sm: 4, lg: 6 }} m={[1, 2, , 4]} />
        );
      }
    `;

    const extracted = extractStylesFromCode(componentCode);
    const usages = extractComponentUsage(usageCode);
    const usageMap = buildUsageMap(usages);

    const generator = new CSSGenerator({ prefix: 'animus' });
    const result = generator.generateFromExtracted(
      extracted[0],
      groupDefinitions,
      testTheme,
      usageMap
    );

    // Verify usage map for responsive values
    expect(usageMap.Card).toBeDefined();
    expect(usageMap.Card.p).toEqual(new Set(['2:_', '4:sm', '6:lg']));
    expect(usageMap.Card.m).toEqual(new Set(['1:_', '2:xs', '4:sm']));

    // Snapshot CSS output
    expect(result.css).toMatchSnapshot();
  });

  it('should handle multiple components with different usage patterns', () => {
    const componentCode = `
      const Button = animus
        .styles({
          cursor: 'pointer',
          border: 'none'
        })
        .groups({ space: true })
        .asElement('button');

      const Text = animus
        .styles({
          lineHeight: 1.5
        })
        .groups({ space: true, typography: true })
        .asElement('p');
    `;

    const usageCode = `
      function Page() {
        return (
          <div>
            <Button p={3} m={0} />
            <Button px={4} py={2} />
            <Text p={2} fontSize="lg" lineHeight={1.8} />
            <Text m={1} fontSize="sm" />
          </div>
        );
      }
    `;

    const extracted = extractStylesFromCode(componentCode);
    const usages = extractComponentUsage(usageCode);
    const usageMap = buildUsageMap(usages);

    const generator = new CSSGenerator({ prefix: 'animus' });

    // Generate CSS for both components
    const buttonCSS = generator.generateFromExtracted(
      extracted[0],
      groupDefinitions,
      testTheme,
      usageMap
    );

    const textCSS = generator.generateFromExtracted(
      extracted[1],
      groupDefinitions,
      testTheme,
      usageMap
    );

    // Verify usage maps for both components
    expect(usageMap.Button).toBeDefined();
    expect(usageMap.Button.p).toEqual(new Set(['3:_']));
    expect(usageMap.Button.m).toEqual(new Set(['0:_']));
    expect(usageMap.Button.px).toEqual(new Set(['4:_']));
    expect(usageMap.Button.py).toEqual(new Set(['2:_']));

    expect(usageMap.Text).toBeDefined();
    expect(usageMap.Text.p).toEqual(new Set(['2:_']));
    expect(usageMap.Text.m).toEqual(new Set(['1:_']));
    expect(usageMap.Text.fontSize).toEqual(new Set(['lg:_', 'sm:_']));
    expect(usageMap.Text.lineHeight).toEqual(new Set(['1.8:_']));

    // Snapshot CSS outputs
    expect({
      button: buttonCSS.css,
      text: textCSS.css,
    }).toMatchSnapshot();
  });

  it('should handle edge cases in usage collection', () => {
    const componentCode = `
      const Flex = animus
        .styles({
          display: 'flex'
        })
        .groups({ space: true })
        .asElement('div');
    `;

    const usageCode = `
      function EdgeCases() {
        const dynamicValue = Math.random() > 0.5 ? 2 : 4;
        
        return (
          <>
            {/* Boolean prop */}
            <Flex p />
            
            {/* String value */}
            <Flex m="auto" />
            
            {/* Number zero */}
            <Flex p={0} />
            
            {/* Null/undefined - should be ignored */}
            <Flex m={null} p={undefined} />
            
            {/* Dynamic value - won't be extracted */}
            <Flex p={dynamicValue} />
            
            {/* Spread props - not handled yet */}
            <Flex {...props} />
          </>
        );
      }
    `;

    const extracted = extractStylesFromCode(componentCode);
    const usages = extractComponentUsage(usageCode);
    const usageMap = buildUsageMap(usages);

    const generator = new CSSGenerator({ prefix: 'animus' });
    const result = generator.generateFromExtracted(
      extracted[0],
      groupDefinitions,
      testTheme,
      usageMap
    );

    // Verify edge case handling
    expect(usageMap.Flex).toBeDefined();
    expect(usageMap.Flex.p).toEqual(new Set(['true:_', '0:_']));
    expect(usageMap.Flex.m).toEqual(new Set(['auto:_', 'null:_']));

    // Verify extracted usages include all cases
    expect(usages).toHaveLength(6); // 6 Flex components

    // Snapshot CSS output only
    expect(result.css).toMatchSnapshot();
  });
});
