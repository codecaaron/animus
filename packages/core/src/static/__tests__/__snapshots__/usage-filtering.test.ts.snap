// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`Usage-based atomic CSS filtering should handle edge cases in usage collection 1`] = `
"/* Base Styles */
.animus-Flex-f4x {
  display: flex;
}

/* Atomic Utilities from Groups */
/* Group: space */
.animus-m-auto {
  margin: auto;
}
.animus-m-null {
  margin: null;
}
.animus-p-true {
  padding: true;
}
.animus-p-0 {
  padding: 0;
}"
`;

exports[`Usage-based atomic CSS filtering should handle multiple components with different usage patterns 1`] = `
{
  "button": "/* Base Styles */
.animus-Button-b6n {
  border: none;
  cursor: pointer;
}

/* Atomic Utilities from Groups */
/* Group: space */
.animus-m-0 {
  margin: 0;
}
.animus-p-3 {
  padding: 16px;
}
.animus-px-4 {
  padding-left: 4;
  padding-right: 4;
}
.animus-py-2 {
  padding-top: 8px;
  padding-bottom: 8px;
}",
  "text": "/* Base Styles */
.animus-Text-t4t {
  line-height: 1.5;
}

/* Atomic Utilities from Groups */
/* Group: space */
.animus-m-1 {
  margin: 4px;
}
.animus-p-2 {
  padding: 8px;
}
/* Group: typography */
.animus-fs-lg {
  font-size: 20px;
}
.animus-fs-sm {
  font-size: 14px;
}",
}
`;

exports[`Usage-based atomic CSS filtering should handle responsive values in usage tracking 1`] = `
"/* Base Styles */
.animus-Card-c4d {
  background-color: var(--animus-colors-white);
  border-radius: 8px;
}

/* Atomic Utilities from Groups */
/* Group: space */
.animus-m-1 {
  margin: 4px;
}
@media screen and (min-width: 480px) {
  .animus-m-2-xs {
    margin: 8px;
  }
}
@media screen and (min-width: 768px) {
  .animus-m-4-sm {
    margin: 4;
  }
}
.animus-p-2 {
  padding: 8px;
}
@media screen and (min-width: 768px) {
  .animus-p-4-sm {
    padding: 4;
  }
}
@media screen and (min-width: 1200px) {
  .animus-p-6-lg {
    padding: 6;
  }
}"
`;

exports[`Usage-based atomic CSS filtering should only generate utilities for actually used prop values 1`] = `
{
  "withUsageFiltering": "/* Base Styles */
.animus-Box-b3x {
  display: block;
  position: relative;
}

/* Atomic Utilities from Groups */
/* Group: space */
.animus-m-0 {
  margin: 0;
}
.animus-mx-2 {
  margin-left: 8px;
  margin-right: 8px;
}
.animus-p-4 {
  padding: 4;
}
.animus-p-1 {
  padding: 4px;
}
@media screen and (min-width: 480px) {
  .animus-p-2-xs {
    padding: 8px;
  }
}",
  "withoutUsageFiltering": "/* Base Styles */
.animus-Box-b3x {
  display: block;
  position: relative;
}

/* Atomic Utilities from Groups */
/* Group: space */",
}
`;
